sdks/python/apache_beam/io/range_trackers.py

```diff
@@ -324,3 +398,89 @@ def set_current_position(self, position):
 
   def fraction_consumed(self):
     return self._range_tracker.fraction_consumed()
+
+
+class LexicographicKeyRangeTracker(OrderedPositionRangeTracker):
+  """
+  A range tracker that tracks progress through a lexicographically
+  ordered keyspace of strings.
+  """
+
+  @classmethod
+  def fraction_to_position(cls, fraction, start=None, end=None):
+    """
+    Linearly interpolates a key that is lexicographically
+    fraction of the way between start and end.
+    """
+    assert 0 <= fraction <= 1, fraction
+    if start is None:
+      start = ''
+    if fraction == 1:
+      return end
+    elif fraction == 0:
+      return start
+    else:
+      if not end:
+        common_prefix_len = len(start) - len(start.lstrip('\xFF'))
+      else:
+        for ix, (s, e) in enumerate(zip(start, end)):
+          if s != e:
+            common_prefix_len = ix
+            break
+        else:
+          common_prefix_len = min(len(start), len(end))
+      # Convert the relative precision of fraction (~53 bits) to an absolute
+      # precision needed to represent values between start and end distinctly.
+      prec = common_prefix_len + int(-math.log(fraction, 256)) + 7
+      istart = cls._string_to_int(start, prec)
+      iend = cls._string_to_int(end, prec) if end else 1 << (prec * 8)
+      ikey = istart + int((iend - istart) * fraction)
+      # Could happen due to rounding.
```

Comment:
Explain why do we need to do this adjustment, even if this happened due to rounding? Do we do it in the Java version?